include/hipSYCL/sycl/libkernel/half.hpp

```diff
@@ -64,56 +68,58 @@ class half {
     return fp16::promote_to_float(_data);
   }
 
-  operator double() const {
-    return fp16::promote_to_double(_data);
-  }
-
-  operator int() const {
-    return static_cast<int>(fp16::promote_to_float(_data));
-  }
-
   HIPSYCL_UNIVERSAL_TARGET
   friend half operator+(const half& a, const half& b) noexcept {
+    fp16::half_storage data;
+    // __hipsycl_backend_switch contains an if statement for sscp pass, so we
+    // cannot write `fp16::half_storage data = __hipsycl_backend_switch(...)`.
     __hipsycl_backend_switch(
-      return fp16::builtin_add(a._data, b._data),
-      return __hipsycl_sscp_half_add(a._data, b._data),
-      return fp16::create(__hadd(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data))),
+      data = fp16::builtin_add(a._data, b._data),
+      data = __hipsycl_sscp_half_add(a._data, b._data),
+      data = fp16::create(__hadd(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data))),
       // HIP uses compiler builtin addition for native _Float16 type
-      return fp16::builtin_add(a._data, b._data),
-      return fp16::builtin_add(a._data, b._data))
+      data = fp16::builtin_add(a._data, b._data),
+      data = fp16::builtin_add(a._data, b._data));
```

Comment:
Couldn't you simplify this along the lines of:
```c++
return detail::create_half(__hipsycl_backend_switch(
  fp16::builtin_add(a._data, b._data),
  ....
));
```

Reply:
Ugh... even if SSCP might in principle work with ternary operator, it probably won't work with nvc++, where we need `if target()` branching.. So you are totally right :(